File: network/udp_multicast_client_minimalist_version.py
#
# Minimalist example of an udp multicast client receiving data for a given car
#
# Note: Please refer to "udp_multicast_client.py" for more details about
#       car data (timestamp decoding, counter for qos...)
#

import mcastsocket      # Used to manage network socket in udp multicast
import racedata         # Contains direct helpers and the class RaceData

### Network part
client = mcastsocket.MCastClient()

### Data part
car_id = 1              # <----- PUT HERE YOUR CAR ID
ofs = racedata.get_car_data_unpack_offset(car_id)

while True:
    # Get data from the udp multicast socket and unpack them
    byte_stream = client.recv()
    if byte_stream is None:
        continue

    # Unpack your car data
    car_status, position, orientation, x, y, race_status, counter, timestamp_ns = racedata.unpack_car_data_and_meta(byte_stream, ofs)

    # Do whatever you want with your car data...
    print("car status {:2}, position {:2}, orientation {:3}°, x {: 6.1f}, y {: 6.1f}, race status {:2}, counter {:6}, {:}".
        format(car_status, position, orientation, x, y, race_status, counter, timestamp_ns))

File: network/udp_multicast_client_minimalist_version.c
#include <stdio.h>

#include "racedata.h"
#include "mcastsocket.h"

int main(void)
{
	// Network part
	int client_fd;
	if (mcastclient_init(&client_fd) == false) {
		printf("Error calling mcastclient_init()\n");
		return 1;
	}

	// Data part
	int car_id = 1;                                 // <----- PUT HERE YOUR CAR ID
	struct RaceData racedata;
	struct Car *car = &racedata.cars[car_id];

	while (1) {
		// Get data from the udp multicast socket and unpack them
		if (!mcastclient_recv(client_fd, &racedata, sizeof(racedata)))
			continue;

		printf("car status %i, position %i, orientation %i°, x %f, y %f, race status %i counter %u, timestamp %lu\n",
				car->status, car->position, car->orientation, car->x, car->y,
				racedata.status, racedata.counter, racedata.timestamp);
	}

	return 0;
}
